### ISSUE TITLE
Mesh shape from custom built AABBTree

### ISSUE THREAD

**Issue by velifaro** [jrouwe/JoltPhysics#339]
Hello. Is it possible to add the ability to create a mesh from a manually built AABBTree to speedup scene loading? Since I'm generating a mesh, I already know the distribution of triangles, so I can easily calculate the active edges without additional search, and it's quite easy to build an AABBTree.
For now I have implemented it the same way as in MeshShape and then creating it via JPH::Shape::sRestoreFromBinaryState:
```cpp
auto [vertices, indices, aabb_tree_root] = create_mesh(...);
JPH::AABBTreeToBuffer<TriangleCodec, NodeCodec> buffer;
buffer.Convert(vertices, &aabb_tree_root, error);
JPH::ShapeResult mesh_result = JPH::Shape::sRestoreFromBinaryState(MeshStream(buffer));
```
But this is an extremely strange method and I also use knowledge about internal implementation of Jolt, which may change.
It also is not as efficient as I would like because of the implementation of sRestoreFromBinaryState, where is byte-by-byte reading from the stream. sRestoreFromBinaryState takes around 20% of mesh creation time.

**Comment by jrouwe** [comment]
Hello, can you explain your use case a bit more? Generally I would expect people to create the shapes in a preprocessing step and save them to a binary stream using SaveBinaryState. At runtime you would load this stream, which is the fastest way of getting the data back in memory. I consider the AABBTree code and the fact that there are active edges an implementation detail that may change, so if I expose it I'm painting myself in a corner. Why is it not possible to do this ahead of time?

**Comment by velifaro** [comment]
Of course I would like to bake the whole scene and use the as much preparation as possible. For static objects and models, this what we used. But our scene is a whole planet and we using cube-to-sphere projections for terrain. As a result, we have to use terrain meshes for physics instead of height fields(tile's axes become non-orthogonal). Due to the huge size of the map procedural generation is used for detailed height and coverage tiles. Unfortunately, we cannot prepare all tiles with required accuracy and save everything to disk, it turns out to be too large. Since I know quite a lot about the tile grid when I bend it from height field to mesh I can optimize building AABBTree and finding active edges quite efficiently.

**Comment by jrouwe** [comment]
I don't immediately see a nice and generic solution (i.e. one that does not depend on AABBTree and the AABBTreeToBuffer class). For now I would suggest adding an extra constructor on MeshShape that takes in the AABBTreeToBuffer class and does the last few lines of the other MeshShape constructor:

```
// Move data to this class
mTree.swap(buffer.GetBuffer());

// Check if we're not exceeding the amount of sub shape id bits
if (GetSubShapeIDBitsRecursive() > SubShapeID::MaxBits)
{
	outResult.SetError("Mesh is too big and exceeds the amount of available sub shape ID bits");
	return;
}

outResult.Set(this);
```

And then I'll think about it some more.

**Comment by velifaro** [comment]
Yes. I also doesn't see solution general solution for that specific case. Maybe it will be good to make MeshShape not a final class and its members protected. After that will be possible inherit it and create class with the properties and constructor what user need.
